### DISCUSSION TITLE
set_basename 相关改进

### DISCUSSION THREAD

**Discussion by MasterBalll** [xmake-io/xmake#3691] [Ideas]
根据文档中 [设置目标文件名](https://xmake.io/#/zh-cn/manual/project_target?id=%e8%ae%be%e7%bd%ae%e7%9b%ae%e6%a0%87%e6%96%87%e4%bb%b6%e5%90%8d) 所述，可以用各种选项来区分目标文件。但除此之外，还希望这些不同的选项配置的构建过程相互独立，也就是当自定义选项在不同值间切换，中间文件都能保留，而不会被覆盖。
所以是否可以改进功能或者添加新的接口，来让相应的 deps & objs & gens 文件保存在 `set_basename()` 指定的路径中而不是原始路径？

**Comment by waruqi** [comment]
? basename 跟路径有什么关系？不懂

**Reply by MasterBalll** [reply]
具体来说，比如 xmake.lua 是这样的
```lua
target("abc")
    set_basename("abc_$(op1)")
    ...
target_end()

option("op1")
option_end()
```
需求是，当要在 `--op1="1"`, `--op2="2"`, …… 之间切来切去构建 & 运行时（嵌入式开发会有这个用途，就是要来回测试多种配置下的运行结果），希望不用重新编译。

basename 倒是跟路径没啥关系啦……

目前是当配置为 `--op1="1"` ，对象文件还是会生成在 .objs/abc 下（.deps 与 .gens 目录也是类似的），如果能按照 basename 的设置生成在 .objs/abc_1 应该能满足这种场景，切换配置就不用重新编译了。

**Reply by waruqi** [reply]
路径，得用 set_objectdir, set_dependir, set_targetdir 改，跟 basename 没啥关系，不要用 set_basename

**Reply by MasterBalll** [reply]
好吧，虽然和想象的不太一样，倒也能达到效果。
